### PR TITLE
fix: [DEL-4672]: Download watcher script in every attempt to restart it

### PR DESCRIPTION
**Pull request by gaurav-jain1**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/36453)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 3 Issues 
* Vulnerability 1 Vulnerabilities 
 * Code issue 2 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/36453)

**Comment by gaurav-jain1** [comment]
trigger messagecheck

File: 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java [hunk 1]
```diff
@@ -1605,13 +1609,16 @@ private void watcherUpgrade(boolean heartbeatTimedOut) {
...
+          () -> performWatcherUpgrade(watcherProcess, multiVersionRestartNeeded, expectedVersion));
     }
   }
 
-  private void performWatcherUpgrade(String watcherProcess, boolean multiVersionRestartNeeded) {
+  private void performWatcherUpgrade(String watcherProcess, boolean multiVersionRestartNeeded, String expectedVersion) {
     synchronized (this) {
       try {
+        // Download watcher script before restarting watcher.
+        downloadRunScriptsForWatcher(expectedVersion);
```
Comment (bharatgoelharness) [inline 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java:1621]: what if scripts failed to download?

**Review by bharatgoelharness — CHANGES_REQUESTED** [review]

**Review by bharatgoelharness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36453)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 1621 at r1](https://reviewable.io/reviews/harness/harness-core/36453#-N9uCA0MEdUrVuAWnU2A:-N9uCA0MEdUrVuAWnU2B:b-l4u41j) ([raw file](https://github.com/harness/harness-core/blob/6a5cc6b169bd79ff100eb75da383e07e07669894/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L1621)):*
> ```Java
>       try {
>         // Download watcher script before restarting watcher.
>         downloadRunScriptsForWatcher(expectedVersion);
> ```

We should check for null response here.
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 1665 at r1](https://reviewable.io/reviews/harness/harness-core/36453#-N9uCGPB2DxXg0z_S7Lx:-N9uCGPB2DxXg0z_S7Ly:b9ly087) ([raw file](https://github.com/harness/harness-core/blob/6a5cc6b169bd79ff100eb75da383e07e07669894/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L1665)):*
> ```Java
> 
>     if (restResponse == null) {
>       log.warn("Received null response from manager.");
> ```

lets change the logging specific to delegate

**Review by Vbhatia20 — APPROVED** [review]
lgtm post suggested changes are done, pls ensure these are tested well and also in QA-FREE/QA as this will not land in PREQA

> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36453#-:-NA1vbd42XqH-vD3-XhX:b-971icr)* status: :shipit: complete! ✓

**Pull request merged**